sycl/include/CL/sycl/queue.hpp

```diff
@@ -874,28 +803,44 @@ class __SYCL_EXPORT queue {
   template <class T>
   friend T detail::createSyclObjFromImpl(decltype(T::impl) ImplObj);
 
+#if __SYCL_USE_FALLBACK_ASSERT
+  friend event detail::submitAssertCapture(queue &, event &, queue *,
+                                           const detail::code_location &);
+#endif
+
   /// A template-free version of submit.
   event submit_impl(function_class<void(handler &)> CGH,
                     const detail::code_location &CodeLoc);
   /// A template-free version of submit.
   event submit_impl(function_class<void(handler &)> CGH, queue secondQueue,
                     const detail::code_location &CodeLoc);
 
+  // Function to postprocess submitted command
+  // Arguments:
+  // bool IsKernel - true if the submitted command was kernel, false otherwise
+  // bool KernelUsesAssert - true if submitted kernel uses assert, only
+  //                         meaningful when IsKernel is true
+  // event &Event - event after which post processing should be executed
+  using SubmitPostProcessF = std::function<void(bool, bool, event &)>;
```

Comment:
Suggest changing the lambda type to take handler and, probably, event objects. Doing so would allow us to extend it's functionality later if needed.

Reply:
**Briefly**: I can't find how it would be possible without storing the required information first.

**Verbose:**
When submitting a kernel, `handler` class moves each and every of its member fields. This makes these fields unusable after `handler::finalize()`. (I think this behaviour shouldn't be changed). At the same time, we want to enqueue smth so, that it'll work after the user's kernel finish i.e. we need an event which represents user's kernel which is available only after `handler::finalize()` does its work.
The only way I see, is to employ this sort of signature: `void(handler &, AdditionalInfo &, Event &)`. The first argument is the handler itself (dunno why this would be needed **after** `handler::finalize()`), the second one is some information stored prior to `handler::finalize()` and the latter argument is the event, which represents submitted kernel.

Reply:
I see...
It is also problematic to define AdditionalInfo type. As an crazy option we could introduce one more lambda - "pre-finalize" which can be called before handler::finalize which a handler arg and capture all required info into a variable known to both lambdas in the header - some var which both lambdas capture.
But, probably, it is overkill.
So, I'm OK with current implementation, it should not be a problem to add a new overload later if there will be need in that.

Reply:
> It is also problematic to define AdditionalInfo type.

It could be something like this:

```c++
class AdditionalInfo {
public:
  enum InfoKind {
    AssertUsedByKernel,
    EnqueuedIsKernel,
  };
  std::variant get(InfoKind) const;
};
```

Though, I can agree that we can add it later when the need arises.